src/libtest/lib.rs

```diff
@@ -876,11 +890,43 @@ fn run_tests<F>(opts: &TestOpts, tests: Vec<TestDescAndFn>, mut callback: F) ->
                 // that hang forever.
                 callback(TeWait(test.desc.clone(), test.testfn.padding()))?;
             }
+            running_tests.insert(test.desc.clone(), Duration::from_secs(TEST_WARN_TIMEOUT_S));
             run_test(opts, !opts.run_tests, test, tx.clone());
             pending += 1;
         }
 
-        let (desc, result, stdout) = rx.recv().unwrap();
+        let mut res;
+        if let Some(min_timeout) = running_tests.values().min().cloned() {
+            loop {
+                let before = Instant::now();
+                res = rx.recv_timeout(min_timeout);
+                let elapsed = Instant::now() - before;
+
+                let mut to_remove = Vec::new();
+                for (desc, time_left) in &mut running_tests {
+                    if *time_left >= elapsed {
+                        *time_left -= elapsed;
+                    } else {
```

Comment:
You may be able to avoid doing these updates if instead of storing the remaining `Duration` in the hash map you instead stored an `Instant` representing the deadline. If that makes sense, do you mind trying to reformulate it like that and seeing if it results in less code?